examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno

```diff
@@ -0,0 +1,443 @@
+// contract recipients are using GRC1155 protocol.
+func (s *basicGRC1155Token) SafeMint(to std.Address, tid TokenID, amount uint64) error {
+	caller := std.GetOrigCaller()
+
+	err := s.mint(to, tid, amount)
+	if err != nil {
+		return err
+	}
+
+	if !s.doSafeTransferAcceptanceCheck(caller, zeroAddress, to, tid, amount) {
+		return ErrTransferToRejectedOrNonGRC1155Receiver
+	}
+
+	return nil
+}
+
+// Batch version of `SafeMint()`. Also checks that
+// contract recipients are using GRC1155 protocol.
+func (s *basicGRC1155Token) SafeBatchMint(to std.Address, batch []TokenID, amounts []uint64) error {
+	caller := std.GetOrigCaller()
+
+	err := s.mintBatch(to, batch, amounts)
+	if err != nil {
+		return err
+	}
+
+	if !s.doSafeBatchTransferAcceptanceCheck(caller, zeroAddress, to, batch, amounts) {
+		return ErrTransferToRejectedOrNonGRC1155Receiver
+	}
+
+	return nil
+}
+
+// Destroys `amount` tokens of token type `id` from `from`.
+func (s *basicGRC1155Token) Burn(from std.Address, tid TokenID, amount uint64) error {
+	err := s.burn(from, tid, amount)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Batch version of `Burn()`
+func (s *basicGRC1155Token) BatchBurn(from std.Address, batch []TokenID, amounts []uint64) error {
+	err := s.burnBatch(from, batch, amounts)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+/* Helper methods */
+
+// Helper for SetApprovalForAll(): approve `operator` to operate on all of `owner` tokens
+func (s *basicGRC1155Token) setApprovalForAll(owner, operator std.Address, approved bool) error {
+	if owner == operator {
+		return ErrApprovalToCurrentOwner
+	}
+
+	key := owner.String() + ":" + operator.String()
+	s.operatorApprovals.Set(key, approved)
+
+	event := ApprovalForAllEvent{owner, operator, approved}
+	emit(&event)
+
+	return nil
+}
+
+// Helper for SafeTransferFrom()
+func (s *basicGRC1155Token) safeTransferFrom(from, to std.Address, tid TokenID, amount uint64) error {
+	if err := isValidAddress(from); err != nil {
+		return ErrInvalidAddress
+	}
+	if err := isValidAddress(to); err != nil {
+		return ErrInvalidAddress
+	}
+	if from == to {
+		return ErrCannotTransferToSelf
+	}
+
+	caller := std.GetOrigCaller()
+	batch := []TokenID{tid}
+	amounts := []uint64{amount}
+	s.beforeTokenTransfer(caller, from, to, batch, amounts)
+
+	fromBalance, err := s.BalanceOf(from, tid)
+	if err != nil {
+		return err
+	}
+	if fromBalance < amount {
+		return ErrInsufficientBalance
+	}
+	toBalance, err := s.BalanceOf(to, tid)
+	if err != nil {
+		return err
+	}
+
+	fromBalance -= amount
+	toBalance += amount
+	fromBalanceKey := string(tid) + ":" + from.String()
+	toBalanceKey := string(tid) + ":" + to.String()
+	s.balances.Set(fromBalanceKey, fromBalance)
+	s.balances.Set(toBalanceKey, toBalance)
+
+	event := TransferSingleEvent{caller, from, to, tid, amount}
+	emit(&event)
+
+	s.afterTokenTransfer(caller, from, to, batch, amounts)
+
+	return nil
+}
```

Comment:
what if, to avoid duplicating logic, this entire function was just:

```go
// called from SafeTransferFrom
s.safeBatchTransferFrom(from, to, []TokenID{tid}, []uint64{amount})
```

and the event emission was pulled out of `safeTransferFrom` and batch version, and were put instead in their exported counterparts? (so each function actually emits the correct event)